### ISSUE TITLE
mask_3d = mask.unsqueeze(dim=2).repeat_interleave(repeats=out.shape[2], dim=2) shape관련 질문

### ISSUE THREAD

**Issue by karlie38** [haven-jeon/KoGPT2-chatbot#3]
안녕하세요. 올려주신 코드를 바탕으로 custom해서 제 테스크에 맞게 모델을 만들고 있습니다.

저 같은 경우는 max_len = 300으로 두고 했는데요, 계속 assertion 에러가 나서요 ㅠㅠ

코드를 보니깐

out = self(token_ids)
print('out', out.shape) -> shape: [1, 300, 50000]

mask_3d = mask.unsqueeze(dim=2).repeat_interleave(repeats=out.shape[2], dim=2)
print('mask_3d', mask_3d.shape) -> shape: [1, 320, 50000]

 mask_out = torch.where(mask_3d == 1, out, self.neg * torch.ones_like(out))

그래서 mask_out에 shape이 안맞아서 에러가 나는데.. 
mask_3d 자체가 out갖고 만드는데, 왜 shape이 320이 나오는지 도통 이해가 안되서요.. 혹시 조언 주시면 감사하겠습니다!!

**Comment by haven-jeon** [comment]
mask shape는 어떻게 되는지요?

**Comment by karlie38** [comment]
안녕하세요!
 
보니깐 mask shape이 안맞아서 그렇네요 ㅜㅜ
 
떄마다 다른데 mask.shpe이 [1, 316], [1, 312] 처럼 max_len = 300 에 맞지 않아서 에러가 나는 것 같아요!
 
그래서, get_item 쪽에서 바꾸려고 했는데.. 왜 안될까요..? 밑에는 제가 custom 해서 코드를 좀 바꾸고 있어요!
 
U_TKN = '<cont>'
S_TKN = '<q>'
BOS = '<s>'
EOS = '</s>'
MASK = '<unused0>'
SENT = '<text>'


class CharDataset(Dataset):
    def __init__(self, chats, tok_path, vocab, max_len=300):
        self._data = chats
        self._tok_path = tok_path
        self.tokenizer = None
        self.first = True
        self.q_token = U_TKN
        self.a_token = S_TKN
        self.sent_token = SENT
        self.bos = BOS
        self.eos = EOS
        self.maskt = MASK
        self.vocab = vocab
        self.max_len = max_len
        self.padder = nlp.data.PadSequence(
            max_len, pad_val=self.vocab[self.vocab.padding_token])

    def _activate_sp(self):
        self.tokenizer = nlp.data.SentencepieceTokenizer(self._tok_path, 0, 0)

    def __len__(self):
        return len(self._data)

    def __getitem__(self, idx):
        if self.tokenizer is None:
            self._activate_sp()
        turn = self._data.iloc[idx]
        q = turn['Q']  # context
        a = turn['A']  # question
        sentiment = str(turn['label'])  # text
        q_toked = [
                      self.q_token,
                  ] + self.tokenizer(q) + [
                      self.eos,
                  ] + [self.sent_token] + self.tokenizer(sentiment) + [
                      self.eos,
                  ]
        q_len = len(q_toked)
        a_toked = [
                      self.a_token,
                  ] + self.tokenizer(a) + [
                      self.eos,
                  ]
        a_len = len(a_toked)
        if q_len + a_len > self.max_len:
            change = q_len + a_len - self.max_len 
            if self.max_len - q_len > 0:
                q_len = q_len - change - 1
                q_toked = q_toked[-q_len:]
                assert q_len == len(q_toked)
        
            elif self.max_len - q_len < 0:
                q_len = q_len - change- 1     
                q_toked = q_toked[-q_len:]
          
                assert q_len == len(q_toked)

        # [mask, mask, ...., mask, ..., <bos>,..A.. <eos>, <pad>....]
        labels = [
                     self.maskt,
                 ] * q_len + a_toked[1:]
        if self.first:
            logging.info("contexts : {}".format(q))
            logging.info("toked ctx: {}".format(q_toked))
            logging.info("response : {}".format(a))
            logging.info("toked response : {}".format(a_toked))
            logging.info('labels {}'.format(labels))
            self.first = False
        mask = [0] * q_len + [1] * a_len + [0] * (self.max_len - q_len - a_len)
        return (self.padder(self.vocab[q_toked + a_toked]), np.array(mask),
                self.padder(self.vocab[labels]))

이 get_item에서 q_len + a_len이 max_len보다 길면 q_len에서 그 차이만큼 적게 토큰이 input으로 들어가도록 하고 있는데.. 왜 안될까요 ㅜㅜ

에러가 나는 mask는 거의 [0, 0, ~ 1, 1] 뒷 부분에 0이 없는 애들( q_len + a_len > max_len)인 경우같은데.. 어떻게 고쳐야할지 작은 조언 좀 주시면 정말 감사하겠습니다!!!

karlie38 드림

-----Original Message-----

**Comment by haven-jeon** [comment]
9faa8aae64498bd068079

**Comment by haven-jeon** [comment]
q_len이 max_len보다 큰 경우에 이슈가 있었네요. 코드 업데이트 했으니 확인 해보세요.
